### DISCUSSION TITLE
Side by side maps corrupted after closing the side bar

### DISCUSSION THREAD

**Discussion by mlandsfeld** [tethysplatform/tethys#755] [Q&A]
Hello,
In my side by side openlayers maps Tethys application, I am able to retrieve feature name and values from the polygons displayed in one of the maps and display those values. But when I close the left hand side bar the maps are resized and the background layer and the overlays match up just fine. But when I click on the polygons to retrieve the feature values, it seems that the mapping of the mouse location is still set to the previous map size. So the retrievals are incorrect. However, if I open and close the debugger, in Chrome, the maps are reset correctly and the retrievals work again. Any ideas on what might be causing this? Thanks!

**Comment by ezrajrice** [comment]
This is likely caused by the "resize" event not being picked up by whatever sets the location mapping. You can test this by triggering the event in your code and seeing if that works. It's just a temporary solution used to debug. The actual fix should include figuring out why this is happening and where in the code the event is not being passed through.

**Comment by mlandsfeld** [comment]
I tried to catch the event in my main.js script like this:

window.onresize = function()
{
	console.log("onresize...");
	map_fcast.updateSize();
	map_eo.updateSize();
};

But I am not getting a message in the console. Any ideas why that might be?
Thanks

**Reply by ezrajrice** [reply]
Sorry, I told you something that was really outdated. The onresize event only occurs when the window is resized. It no longer supports that event on any DOM element. You'll need to use the MutationObserver to achieve this. A little digging in the inspection shows that the class changes on the 'app-content-wrapper' div when the navigation pane is opened or closed. This is the change or mutation that we will attempt to capture. This example was adapted from the [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver). You're looking for the "show-nav" class to change. Try

```javascript
// Select the node that will be observed for mutations
const targetNode = document.getElementById('app-content-wrapper');

// Options for the observer (which mutations to observe)
const config = { attributes: true };

// Optionally collect the previous class state
let targetNodeClasses = targetNode.classList;

// Callback function to execute when mutations are observed
const callback = function(mutationsList, observer) {
    // Use traditional 'for loops' for IE 11
    for(const mutation of mutationsList) {
        if (mutation.attributeName === 'class' & targetNodeClasses != targetNode.classList) {
            targetNodeClasses = targetNode.classList;
            console.log("Nav pane visibility changed");
            map_fcast.updateSize();
            map_eo.updateSize()
        }
    }
};

// Create an observer instance linked to the callback function
const observer = new MutationObserver(callback);

// Start observing the target node for configured mutations
observer.observe(targetNode, config);

// Later, you can stop observing
observer.disconnect();

```
EDIT: This had a typo on the if statement so I fixed it. You should still refer to the later post for a better answer that has been more fully tested.

**Reply by mlandsfeld** [reply]
I set my target to this div in my home.html file:

const targetNode = document.getElementById('inner-app-content');

Is this a legitimate target? 
When I load the page, I get an "Uncaught SyntaxError: Unexpected identifier" at the if statement. When I take this part out:

mutation.attributeName === 'class'

I don't get the error but it looks like I am not getting into the callback function at all.  Any suggestions? Below is a listing of the code.


// Select the node that will be observed for mutations
const targetNode = document.getElementById('inner-app-content');

// Options for the observer (which mutations to observe)
const config = { attributes: true };

// Optionally collect the previous class state
let targetNodeClasses = targetNode.classList;

// Callback function to execute when mutations are observed
const callback = function(mutationsList, observer) {
    console.log("Callback...");
    // Use traditional 'for loops' for IE 11
    for(const mutation of mutationsList) {
        if (targetNodeClasses != targetNode.classList) {
            targetNodeClasses = targetNode.classList;
            console.log("Nav pane visibility changed", mutation);
            map_fcast.updateSize();
            map_eo.updateSize();
        }
    }
};

// Create an observer instance linked to the callback function
const observer = new MutationObserver(callback);

// Start observing the target node for configured mutations
observer.observe(targetNode, config);

**Comment by ezrajrice** [comment]
Use this. I just finished testing it and had it working without issue. If you're still having trouble then we can setup a quick call or something to get you sorted.

```javascript
    // Select the node that will be observed for mutations
    const targetNode = document.getElementById('app-content-wrapper');

    // Options for the observer (which mutations to observe)
    const config = { attributes: true };

    // Optionally collect the previous class state
    let showingNav = targetNode.classList.contains('show-nav');

    // Callback function to execute when mutations are observed
    const callback = function(mutationsList, observer) {
        // Use traditional 'for loops' for IE 11
        for(const mutation of mutationsList) {
            if (mutation.attributeName === 'class' && mutation.target.classList.contains('show-nav') != showingNav) {
                showingNav = mutation.target.classList.contains('show-nav');
                console.log("Show nav pane: " + showingNav);
                map_fcast.updateSize();
                map_eo.updateSize();
            }
        }
    };

    // Create an observer instance linked to the callback function
    const observer = new MutationObserver(callback);

    // Start observing the target node for configured mutations
    observer.observe(targetNode, config);
```

**Reply by mlandsfeld** [reply]
Hi Ezra,
I implemented this but I am still not executing anything in the callback function. Maybe a quick call or video meeting could help sort this out. Thanks!

**Reply by ezrajrice** [reply]
> Hi Ezra, I implemented this but I am still not executing anything in the callback function. Maybe a quick call or video meeting could help sort this out. Thanks!

Send me a meeting link at REDACTED and we'll get this sorted.

EDIT:
The call resulted in this looking like an OpenLayers issue. There are a few moving parts. The above code correctly catches when the nav pane is opened or closed, but the element id is different based on the Tethys version the app was created in. When the nav pane is opened or closed, it is being shifted off screen. This causes the visible inner app content to stretch to fill the space. When this occurs, OpenLayers does not recognize that the map div has changed sizes. Catching the div size change and doing `map.updateSize([div_width, div_height])` doesn't seem to work. The OpenLayers map layers seems to be locked in on the size that the map is originally rendered in which includes the nav pane being visible.

**Comment by swainn** [comment]
Thanks for jumping on this @ezrajrice. The logic that we have in the tethys map view gizmo that seems to work well is the following:

```javascript
//function to change size of the map when the map element size changes
$map_element.changeSize(function($this){
  m_map.updateSize();
});
```

You can find it here: https://github.com/tethysplatform/tethys/blob/master/tethys_gizmos/static/tethys_gizmos/js/tethys_map_view.js#L1064-L1067

It looks like OpenLayers has a built-in `changeSize()` method. Have you given that a try?

**Reply by mlandsfeld** [reply]
We did try changeSize function and it did not work for us. And that was the case until I wrapped it in a setTimeout() function and gave it a 400 ms delay. I can see what is happening, the navigation bar is animated, when opening and closing, and as it is resizing, you can see the maps resize a few times during the change. The delay might not be the best solution. I am thinking that removing the animation effect would produce more reliable results or maybe catch an event message that is sent when the animation is done. Or is there something else that would be better? Ezra helped me set up the MutationObserver code listed a couple posts back. I am still using that pattern.
Thanks for checking in on this!
